src/node.rs

```diff
@@ -0,0 +1,441 @@
+
+    #[inline]
+    fn get_undetermined_events(&self, round: usize) -> Vec<(usize, EventHash)> {
+        let next_consensus = self.get_next_consensus();
+        (next_consensus..round)
+            .filter(|r| !self.consensus.contains(r))
+            .map(|r| get_round_pairs(&self.rounds[r]).into_iter())
+            .flatten()
+            .filter(|(_,h)| self.hashgraph.get(&h).unwrap().is_undefined())
+            .collect::<Vec<(usize, EventHash)>>()
+    }
+
+    #[inline]
+    fn get_round_witnesses(
+        &self,
+        round: usize,
+        hash: &EventHash
+    ) -> Result<BTreeSet<EventHash>, Error> {
+        let mut hits: HashMap<PeerId, usize> = HashMap::new();
+        let event = self.hashgraph.get(hash)?;
+        let prev_round = round - 1;
+        for (creator, event_hash) in event.can_see().iter() {
+            let possible_witness = self.hashgraph.get(event_hash)?;
+            if possible_witness.round()? == prev_round {
+                for (_creator, _event_hash) in possible_witness.can_see().iter() {
+                    let r = self.hashgraph.get(_event_hash)?.round()?;
+                    if r == prev_round {
+                        let new_val = hits.get(creator).map(|v| *v+1).unwrap_or(1);
+                        hits.insert(creator.clone(), new_val);
+                    }
+                }
+            }
+        }
+        let r = &self.rounds[prev_round];
+        let map_iter = hits.into_iter()
+            .filter(|(_,v)| *v > self.super_majority)
+            .map(|(c, _)| r.witnesses_map()[&c].clone());
+        Ok(BTreeSet::from_iter(map_iter))
+    }
+
+    #[inline]
+    fn get_voters(&self) -> Vec<(usize, EventHash)> {
+        let next_consensus = self.get_next_consensus();
+        self.rounds[next_consensus..self.rounds.len()].iter()
+            .flat_map(|r| get_round_pairs(r))
+            .collect()
+    }
+
+    #[inline]
+    fn get_next_consensus(&self) -> usize {
+        self.consensus.iter()
+            .last()
+            .map(|v| *v + 1)
+            .unwrap_or(0)
+    }
+
+    #[inline]
+    fn set_event_can_see_self(&mut self, hash: &EventHash) -> Result<(), Error> {
+        let event = self.hashgraph.get_mut(&hash)?;
+        let creator = event.creator().clone();
+        event.add_can_see(creator, hash.clone());
+        Ok(())
+    }
+
+    #[inline]
+    fn assign_round(&mut self, hash: &EventHash) -> Result<usize, Error> {
+        if self.hashgraph.get(hash)?.is_root() {
+            assign_root_round(self.hashgraph.get_mut(&hash)?)
+        } else {
+            self.assign_non_root_round(hash)
+        }
+    }
+
+    #[inline]
+    fn assign_non_root_round(&mut self, hash: &EventHash) -> Result<usize, Error> {
+        let events_parents_can_see = self.hashgraph.events_parents_can_see(hash)?;
+        let mut r = self.get_parents_round(hash)?;
+        let mut hits: HashMap<PeerId, usize> = HashMap::new();
+        for (_, h) in events_parents_can_see.iter() {
```

Comment:
Can we use rayon here?

Reply:
TBH, I didn't know about `rayon` and it looks lovely. Yes, we could (and I'd argue we should!). https://github.com/Fantom-foundation/lachesis-rs/issues/8